[llvm-cov] Fix branch counts of template functions (#111743)

stma247 opened the pull request:
Added option "--unify-instantiations" to llvm-cov export to combine branch execution counts of C++ template instantiations.

on-behalf-of: @e-solutions-GmbH <info@esolutions.de>

stma247 commented:
Hi @evodius96, could you please review my changes.

stma247 commented:
Ping

evodius96 commented:
Sorry for the delay, I was out of town for a family emergency. I will take a look at this this week! Thanks

evodius96 commented on llvm/tools/llvm-cov/CodeCoverage.cpp:
```diff
@@ -1270,13 +1270,18 @@ int CodeCoverageTool::doExport(int argc, const char **argv,
 
+  cl::opt<bool> UnifyInstantiations("unify-instantiations", cl::Optional,
+                                    cl::desc("Unify function instantiations"),
+                                    cl::init(false), cl::cat(ExportCategory));
```
Would it be worthwhile to turn this ON by default?  Then the summary would be more consistent with the summary output of `llvm-cov show`.

stma247 replied:
Okay, I changed the default to true.

evodius96 commented on llvm/tools/llvm-cov/CoverageExporterLcov.cpp:
```diff
@@ -75,70 +90,120 @@ void renderLineExecutionCounts(raw_ostream &OS,
-    if (ViewDepth == 0)
-      SrcLine = Expansion.Region.LineStart;
+    // Track the path to the nested expansions
+    NestedPath.push_back(Expansion.Region.startLoc());
```
Much better than tracking LineStart. Thanks!

stma247 replied:
Yes, it's required to avoid unifying different expanded branches of the same root line.

evodius96 commented on llvm/tools/llvm-cov/CoverageExporterLcov.cpp:
```diff
@@ -75,70 +90,120 @@ void renderLineExecutionCounts(raw_ostream &OS,
+           SumBranch->NestedPath == NextBranch->NestedPath) {
+      SumBranch->ExecutionCount += NextBranch->ExecutionCount;
+      SumBranch->FalseExecutionCount += NextBranch->FalseExecutionCount;
+      NextBranch->TrueFolded = true;
```
OK. So you're summing the counts and then hiding the additional `NextBranch` by marking it as folded.  Arguably, you're overloading the definition of what `Folded` means here, but I'm OK with it. Can you add a brief comment into the code indicating why you're doing this to make it clear what the expected effect is?

stma247 replied:
Okay, I added a new flag `Ignore` to `struct NestedCountedRegion` to make it more clear.

evodius96 commented on llvm/tools/llvm-cov/CoverageExporterLcov.cpp:
```diff
@@ -43,9 +43,26 @@
+  // Contains the path to default and expanded branches
+  // Size is 1 for default branches and greater 1 for expanded branches.
+  std::vector<LineColPair> NestedPath;
+  // Indicates whether this item should at rendering
```
Nit: For LLVM, it's typically to end comments with a period.

stma247 replied:
Ah, I didn't know that and updated all my added comments now.

evodius96 reviewed (approved):
Other than the nit about comments, LGTM. Thanks!

stma247 commented:
Thanks for the review.
When will this PR be merged?

evodius96 commented:
> Thanks for the review.
> 
> When will this PR be merged?

You can trigger the merge yourself from this pull request.

stma247 committed:
[llvm-cov] Fix branch counts of template functions (#111743)

Added option "--unify-instantiations" to llvm-cov export to combine
branch execution counts of C++ template instantiations.

on-behalf-of: @e-solutions-GmbH <info@esolutions.de>

stma247 commented:
> > Thanks for the review.
> > When will this PR be merged?
> 
> You can trigger the merge yourself from this pull request.

The merge button is always disabled for me with info:
Only those with [write access] to this repository can merge pull requests.

stma247 commented:
Ping

evodius96 commented:
> The merge button is always disabled for me with info:
> Only those with [write access] to this repository can merge pull requests.

Apologies -- last week was a holiday in the US.  Ah, I see. I think you should request write access to the repo, but the process has changed since I did it.  Nevertheless, I will attempt to merge this for you since I have the ability to do that.

evodius96 commented:
@stma247  Looks like the LIT test failed on a couple of the build bots.  I reverted the commit until that can be addressed.

stma247 commented:
The new PR is #135074
